pkg/kubelet/network/dns/dns_windows.go

```diff
@@ -0,0 +1,182 @@
+	CurrentDNSServer *syscall.IpAddrString
+	DNSServerList    syscall.IpAddrString
+	NodeType         uint32
+	ScopeID          [maxScopeIDLen + 4]byte
+	EnableRouting    uint32
+	EnableProxy      uint32
+	EnableDNS        uint32
+}
+
+var (
+	// GetNetworkParams can be found in iphlpapi.dll
+	// see: https://docs.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getnetworkparams?redirectedfrom=MSDN
+	iphlpapidll          = windows.MustLoadDLL("iphlpapi.dll")
+	procGetNetworkParams = iphlpapidll.MustFindProc("GetNetworkParams")
+)
+
+func getHostDNSConfig(resolverConfig string) (*runtimeapi.DNSConfig, error) {
+	var (
+		hostDNS, hostSearch []string
+		err                 error
+	)
+	// Get host DNS settings
+	if resolverConfig != "" {
+		hostDNS, err = getDNSServerList()
+		if err != nil {
+			klog.ErrorS(err, "Could not get the host's DNS Server List")
+			return nil, err
+		}
+		hostSearch, err = getDNSSuffixList()
+		if err != nil {
+			klog.ErrorS(err, "Could not get the host's DNS Suffix List")
+			return nil, err
+		}
+	}
+	return &runtimeapi.DNSConfig{
+		Servers:  hostDNS,
+		Searches: hostSearch,
+	}, nil
+}
+
+func elemInList(elem string, list []string) bool {
+	for _, e := range list {
+		if e == elem {
+			return true
+		}
+	}
+	return false
+}
+
+func getRegistryValue(reg, key string) string {
+	regKey, err := registry.OpenKey(registry.LOCAL_MACHINE, reg, registry.QUERY_VALUE)
+	if err != nil {
+		return ""
+	}
+	defer regKey.Close()
+
+	regValue, _, err := regKey.GetStringValue(key)
+	if err != nil {
+		return ""
+	}
+	return regValue
+}
+
+// getDNSSuffixList reads DNS config file and returns the list of configured DNS suffixes
+func getDNSSuffixList() ([]string, error) {
+	// We start with the general suffix list that apply to all network connections.
+	allSuffixes := []string{}
+	suffixes := getRegistryValue(netRegistry, "SearchList")
+	if suffixes != "" {
+		allSuffixes = strings.Split(suffixes, ",")
+	}
+
+	// Then we append the network-specific DNS suffix lists.
+	regKey, err := registry.OpenKey(registry.LOCAL_MACHINE, netIfacesRegistry, registry.ENUMERATE_SUB_KEYS)
+	if err != nil {
+		return nil, err
+	}
+	defer regKey.Close()
+
+	ifaces, err := regKey.ReadSubKeyNames(0)
+	if err != nil {
+		return nil, err
+	}
+	for _, iface := range ifaces {
+		suffixes := getRegistryValue(fmt.Sprintf("%s\\%s", netIfacesRegistry, iface), "SearchList")
+		if suffixes == "" {
+			continue
+		}
+		for _, suffix := range strings.Split(suffixes, ",") {
+			if !elemInList(suffix, allSuffixes) {
+				allSuffixes = append(allSuffixes, suffix)
+			}
+		}
+	}
+
+	return allSuffixes, nil
+}
+
+func getNetworkParams() (*FixedInfo, error) {
+	// We don't know how big we should make the byte buffer, but the call will tell us by
+	// setting the size afterwards.
+	var size int
+	buffer := make([]byte, 1)
+	procGetNetworkParams.Call(
+		uintptr(unsafe.Pointer(&buffer[0])),
+		uintptr(unsafe.Pointer(&size)),
+	)
+
+	buffer = make([]byte, size)
```

Comment:
it would be best to check the return value of the call to be `ERROR_BUFFER_OVERFLOW `. Otherwise the `size` may be undefined (I assume) and can be negative resulting in panic here

Reply:
Done